scipy/spatial/transform/_rigid_transformation.pyx

```diff
@@ -0,0 +1,1978 @@
+# cython: cpow=True
+
+import numpy as np
+from scipy._lib._util import check_random_state, _transition_to_rng
+from ._rotation import Rotation, compose_quat
+
+cimport numpy as np
+cimport cython
+
+np.import_array()
+
+
+cdef _create_skew_matrix(vec):
+    """Create skew-symmetric (aka cross-product) matrix for stack of vectors."""
+    result = np.zeros((len(vec), 3, 3))
+    result[:, 0, 1] = -vec[:, 2]
+    result[:, 0, 2] = vec[:, 1]
+    result[:, 1, 0] = vec[:, 2]
+    result[:, 1, 2] = -vec[:, 0]
+    result[:, 2, 0] = -vec[:, 1]
+    result[:, 2, 1] = vec[:, 0]
+    return result
+
+
+cdef _compute_se3_exp_translation_transform(rot_vec):
+    """Compute transform matrix from se3 translation part to SE3 translation.
+    
+    The transform matrix depends on the rotation vector.
+    """
+    angle = np.linalg.norm(rot_vec, axis=1)
+    mask = angle < 1e-3
+
+    k1 = np.empty(len(rot_vec))
+    k1[mask] = 0.5 - angle[mask]**2 / 24 + angle[mask]**4 / 720
+    k1[~mask] = (1.0 - np.cos(angle[~mask])) / angle[~mask]**2
+
+    k2 = np.empty(len(rot_vec))
+    k2[mask] = 1/6 - angle[mask]**2 / 120 + angle[mask]**4 / 5040
+    k2[~mask] = (angle[~mask] - np.sin(angle[~mask])) / angle[~mask]**3
+
+    s = _create_skew_matrix(rot_vec)
+
+    return np.identity(3) + k1[:, None, None] * s + k2[:, None, None] * s @ s
+
+
+cdef _compute_se3_log_translation_transform(rot_vec):
+    """Compute transform matrix from SE3 translation to se3 translation part.
+    
+    It is inverse of `_compute_se3_exp_translation_transform` in a closed analytical
+    form.
+    """
+    angle = np.linalg.norm(rot_vec, axis=1)
+    mask = angle < 1e-3
+
+    k = np.empty(len(rot_vec))
+    k[mask] = 1/12 + angle[mask]**2 / 720 + angle[mask]**4 / 30240
+    k[~mask] = (1 - 0.5 * angle[~mask] / np.tan(0.5 * angle[~mask])) / angle[~mask]**2
+
+    s = _create_skew_matrix(rot_vec)
+
+    return np.identity(3) - 0.5 * s + k[:, None, None] * s @ s
+
+
+cdef _quaternion_conjugate(quat):
+    conjugate = np.copy(quat)
+    conjugate[:, :3] = -conjugate[:, :3]
+    return conjugate
+
+
+cdef _compose_dual_quaternion(real_part1, dual_part1, real_part2, dual_part2):
+    prod_real = compose_quat(real_part1, real_part2)
+    prod_dual = (np.asarray(compose_quat(real_part1, dual_part2))
+                 + np.asarray(compose_quat(dual_part1, real_part2)))
```

Comment:
`asarray` looks surprising here, is it necessary?

Reply:
The result of compose_quat does not support the + operator. I don't know if there is any better solution.